tests/core/test_otel_logger.py

```diff
@@ -0,0 +1,156 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+from unittest import mock
+
+import pytest
+from opentelemetry.metrics import MeterProvider
+
+from airflow.metrics.otel_logger import (
+    METRIC_NAME_PREFIX,
+    OTEL_NAME_MAX_LENGTH,
+    UP_DOWN_COUNTERS,
+    SafeOtelLogger,
+    _is_up_down_counter,
+)
+
+
+def full_name(name: str):
+    return f"{METRIC_NAME_PREFIX}{name}"
+
+
+@pytest.fixture
+def name():
+    return "test_stats_run"
+
+
+class TestOtelMetrics:
+    def setup_method(self):
+        self.meter = mock.Mock(MeterProvider)
+        self.stats = SafeOtelLogger(otel_provider=self.meter)
+        self.map = self.stats.metrics_map.map
+
+    def test_is_up_down_counter_positive(self):
+        udc = next(iter(UP_DOWN_COUNTERS))
+
+        assert _is_up_down_counter(udc)
+
+    def test_is_up_down_counter_negative(self):
+        assert not _is_up_down_counter("this_is_not_a_udc")
+
+    def test_stat_name_must_be_a_string(self):
+        self.stats.incr(42)
```

Comment:
Shouldn't this and the other incorrect name cases throw an exception?

Reply:
Interesting... it isn't caught as an exception because it just gets folded into an f-string [here](https://github.com/apache/airflow/pull/30873/files#diff-bc07cf67d2d211678aaefd4d99b1e8c6e2c6e26a6987e85c5397ccc78fed47feR89) but the incr() parameter is type-hinted as a string.  I could throw an `if not isInstance(name, str): raise ..." but I wonder if a "validate_name()" helper method that asserts it is a string and truncates as needed might be a better answer?

Reply:
Thinking on this more, SHOULD they?  Do we really want Airflow blowing up because someone implemented a metric with a bad name?  Maybe this should be logged as a warning?

I'll add a more thorough name screening/validation, but maybe an exception isn't the right failure behavior here.

Reply:
Also of note is that the name MAY contain digits, the [OTel spec](https://opentelemetry.io/docs/reference/specification/metrics/api/#instrument-name-syntax) says it must START with an alpha and subsequent characters must belong to the alphanumeric characters, ‘_’, ‘.’, and ‘-’.   Since we are prefixing `airflow.` to the names, then technically the name "42" (as a string) here would qualify as a legal name since it is later getting folded into an f-string, which confuses things.   I will run the validation checks assuming there is no prefix in case we decide to drop that at some point for some reason.

Reply:
> Thinking on this more, SHOULD they? Do we really want Airflow blowing up because someone implemented a metric with a bad name? Maybe this should be logged as a warning?
> 
> I'll add a more thorough name screening/validation, but maybe an exception isn't the right failure behavior here.

We definitely do not want Airflow to blow up, simply because of a bad metric naming. What we should do is to block those `illegal` metric names from getting to the metric registry, and simply posting warning on the airflow log to state that the following metric `XYZ` is illegal (does not start with alphabet) and therefore it was not included.

Reply:
> 
I do believe, however, that the default prefix `airflow` can be changed, is it not? In that case, one can set the prefix to `42` and that might break the OTEL spec..?

Reply:
What I have now (tests are running locally, will push this evening) is a check which validates the name by itself, validates the optional prefix if it's included, and then checks the length of the combined `prefix.name`.  So it should cover us.  Also added unit tests for each of those.